### ISSUE TITLE
Alarm Clock Control

### ISSUE THREAD

**Issue by milymat** [bencevans/node-sonos#160]
Hi,

there are no instructions how to use the AlarmClock in API.md

Maybe someone can post instructions for Create, Update and Delete Alarm.

thx

**Comment by toddanglin** [comment]
I'm far, far from an expert on this library, but I've been spending some time with it this week, so I'll share what I think I know.

**TL;DR** The AlarmClock operations are not "fully" supported in this library yet. But with a little work, you could probably enable them.

There are a number of "stubs" for additional Sonos capabilities in the `lib > services` directory that are not yet exposed via the primary `sonos.js` class. In fact, the only "service" currently being used within `sonos.js` is `ContentDirectory` to support searching for Sonos content.

In _theory_, to use the AlarmClock service without changing `sonos.js`, you'd start by doing something like this:

```javascript
var sonos = new Sonos(ipAddress, port);
var alarmClock = new Services.AlarmClock(sonos.host, sonos.port);
return alarmClock.ListAlarms(opts, function (err, data) {
  // Process response
}
```

What appears to be "missing" that you'd have to figure-out independently:

1. What `options` do the Sonos AlarmClock service methods expect?
2. What `data` do the Sonos AlarmClock methods return?

The stubs that exist today will handle the network request and XML response parsing, but additional work is required to make the methods fully functional.

I think. Again, I'll defer to the experts, but that's my takeaway from the source. Hope it helps.

**Comment by ebaauw** [comment]
>What appears to be "missing" that you'd have to figure-out independently:
>1. What options do the Sonos AlarmClock service methods expect?
>2. What data do the Sonos AlarmClock methods return?

This is actually documented on the ZonePlayer itself.  See `http://<zoneplayer_ip>:1400/xml/AlarmClock1.xml`.

**Comment by bencevans** [comment]
@toddanglin and @ebaauw bossing it 🎉

I'm closing this issue as the service classes mentioned are for advanced users to play with but there's no support for them. They were something I generated a long time ago when thinking of changing the architecture but haven't really been utilised by the library itself. They're in use by some but for easy functionality you'll want to use the documented API calls.

**Comment by bencevans** [comment]
Actually, let's update this issue to be for the AlarmClock control functionality in the main API...

**Comment by svrooij** [comment]
Check out my initial work on this https://github.com/svrooij/node-sonos/tree/alarm
You can now do the following
```node
var alarmService = sonos.alarmClockService()
alarmService.ListAlarms((err, res) => {
  if (err) {
    console.log('Error loading alarms from %s %j', sonos.host, err)
    return
  }
  console.log('Got alarms %s', JSON.stringify(res, null, 2))
})
```

And that returns the following:
```json
{
  "xmlns:u": "urn:schemas-upnp-org:service:AlarmClock:1",
  "CurrentAlarmList": [
    {
      "ID": "76",
      "StartTime": "07:32:00",
      "Duration": "00:15:00",
      "Recurrence": "WEEKDAYS",
      "Enabled": "0",
      "RoomUUID": "RINCON_B*****01400",
      "ProgramURI": "x-rincon-cpcontainer:1006206cspotify%3auser%3aspotify%3aplaylist%3a37i9dQZF1DX6VdMW310YC7?sid=9&flags=8300&sn=7",
      "PlayMode": "SHUFFLE",
      "Volume": "6",
      "IncludeLinkedZones": "0"
    },
    {
      "ID": "81",
      "StartTime": "09:00:00",
      "Duration": "00:15:00",
      "Recurrence": "ONCE",
      "Enabled": "0",
      "RoomUUID": "RINCON_B*********01400",
      "ProgramURI": "x-rincon-cpcontainer:1006206cspotify%3auser%3adigsternl%3aplaylist%3a7g0V4gerBRwKlNuUcqqrg0?sid=9&flags=8300&sn=7",
      "PlayMode": "SHUFFLE",
      "Volume": "8",
      "IncludeLinkedZones": "1"
    }
  ],
  "CurrentAlarmListVersion": "RINCON_0*********01400:1259"
}
```
